synapse/storage/databases/main/cache.py

```diff
@@ -178,7 +183,11 @@ def process_replication_rows(
                     members_changed = set(row.keys[1:])
                     self._invalidate_state_caches(room_id, members_changed)
                 else:
-                    self._attempt_to_invalidate_cache(row.cache_func, row.keys)
+                    res = self._attempt_to_invalidate_cache(row.cache_func, row.keys)
+                    if not res:
+                        external_func = self.external_cached_functions[row.cache_func]
+                        if external_func:
+                            external_func.invalidate(row.keys)
```

Comment:
@reivilibre hum after looking again, it seems like other more specific invalidate functions call `_attempt_to_invalidate_cache` too. It should be safe since I am prepending `__module__` to the function name to avoid clashing with the base code, but it would still be called and degraded performance a bit for those, for not a lot of gain in the end so not sure.

Reply:
In the end I did what you suggest, it's way more clean. Thanks.